File: nanopdf-js/src/geometry.ts
/**
 * Geometry primitives - Point, Rect, Matrix, Quad
 */

import { native_addon } from './native.js';

/**
 * Point-like object
 */
export interface PointLike {
  readonly x: number;
  readonly y: number;
}

/**
 * Rect-like object
 */
export interface RectLike {
  readonly x0: number;
  readonly y0: number;
  readonly x1: number;
  readonly y1: number;
}

/**
 * Matrix-like object
 */
export interface MatrixLike {
  readonly a: number;
  readonly b: number;
  readonly c: number;
  readonly d: number;
  readonly e: number;
  readonly f: number;
}

/**
 * Quad-like object
 */
export interface QuadLike {
  readonly ul: PointLike;
  readonly ur: PointLike;
  readonly ll: PointLike;
  readonly lr: PointLike;
}

/**
 * A 2D point
 */
export class Point implements PointLike {
  readonly x: number;
  readonly y: number;

  constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }

  /**
   * Origin point (0, 0)
   */
  static readonly ORIGIN = new Point(0, 0);

  /**
   * Create a point from a point-like object
   */
  static from(p: PointLike): Point {
    if (p instanceof Point) {
      return p;
    }
    return new Point(p.x, p.y);
  }

  /**
   * Transform this point by a matrix
   */
  transform(m: MatrixLike): Point {
    const result = native_addon.transformPoint(
      { x: this.x, y: this.y },
      { a: m.a, b: m.b, c: m.c, d: m.d, e: m.e, f: m.f }
    );
    return new Point(result.x, result.y);
  }

  /**
   * Calculate distance to another point
   */
  distanceTo(other: PointLike): number {
    const dx = this.x - other.x;
    const dy = this.y - other.y;
    return Math.sqrt(dx * dx + dy * dy);
  }

  /**
   * Add another point
   */
  add(other: PointLike): Point {
    return new Point(this.x + other.x, this.y + other.y);
  }

  /**
   * Subtract another point
   */
  subtract(other: PointLike): Point {
    return new Point(this.x - other.x, this.y - other.y);
  }

  /**
   * Scale by a factor
   */
  scale(factor: number): Point {
    return new Point(this.x * factor, this.y * factor);
  }

  /**
   * Check equality
   */
  equals(other: PointLike): boolean {
    return this.x === other.x && this.y === other.y;
  }

  toString(): string {
    return `Point(${this.x}, ${this.y})`;
  }
}

/**
 * A rectangle defined by two corner points
 */
export class Rect implements RectLike {
  readonly x0: number;
  readonly y0: number;
  readonly x1: number;
  readonly y1: number;

  constructor(x0: number, y0: number, x1: number, y1: number) {
    this.x0 = x0;
    this.y0 = y0;
    this.x1 = x1;
    this.y1 = y1;
  }

  /**
   * Empty rectangle
   */
  static readonly EMPTY = new Rect(Infinity, Infinity, -Infinity, -Infinity);

  /**
   * Infinite rectangle
   */
  static readonly INFINITE = new Rect(-Infinity, -Infinity, Infinity, Infinity);

  /**
   * Unit rectangle (0,0) to (1,1)
   */
  static readonly UNIT = new Rect(0, 0, 1, 1);

  /**
   * Create a rect from a rect-like object
   */
  static from(r: RectLike): Rect {
    if (r instanceof Rect) {
      return r;
    }
    return new Rect(r.x0, r.y0, r.x1, r.y1);
  }

  /**
   * Create a rect from position and size
   */
  static fromXYWH(x: number, y: number, width: number, height: number): Rect {
    return new Rect(x, y, x + width, y + height);
  }

  /**
   * Width of the rectangle
   */
  get width(): number {
    return this.x1 - this.x0;
  }

  /**
   * Height of the rectangle
   */
  get height(): number {
    return this.y1 - this.y0;
  }

  /**
   * Check if the rectangle is empty
   */
  get isEmpty(): boolean {
    return native_addon.isRectEmpty({ x0: this.x0, y0: this.y0, x1: this.x1, y1: this.y1 });
  }

  /**
   * Check if the rectangle is infinite
   */
  get isInfinite(): boolean {
    return this.x0 === -Infinity;
  }

  /**
   * Check if a point is inside the rectangle
   */
  contains(p: PointLike): boolean;
  contains(x: number, y: number): boolean;
  contains(xOrPoint: number | PointLike, y?: number): boolean {
    const px = typeof xOrPoint === 'number' ? xOrPoint : xOrPoint.x;
    const py = typeof xOrPoint === 'number' ? y! : xOrPoint.y;
    return px >= this.x0 && px < this.x1 && py >= this.y0 && py < this.y1;
  }

  /**
   * Union with another rectangle
   */
  union(other: RectLike): Rect {
    const result = native_addon.rectUnion(
      { x0: this.x0, y0: this.y0, x1: this.x1, y1: this.y1 },
      { x0: other.x0, y0: other.y0, x1: other.x1, y1: other.y1 }
    );
    return new Rect(result.x0, result.y0, result.x1, result.y1);
  }

  /**
   * Intersection with another rectangle
   */
  intersect(other: RectLike): Rect {
    const result = native_addon.rectIntersect(
      { x0: this.x0, y0: this.y0, x1: this.x1, y1: this.y1 },
      { x0: other.x0, y0: other.y0, x1: other.x1, y1: other.y1 }
    );
    return new Rect(result.x0, result.y0, result.x1, result.y1);
  }

  /**
   * Expand by including a point
   */
  includePoint(p: PointLike): Rect {
    return new Rect(
      Math.min(this.x0, p.x),
      Math.min(this.y0, p.y),
      Math.max(this.x1, p.x),
      Math.max(this.y1, p.y)
    );
  }

  /**
   * Translate by offset
   */
  translate(dx: number, dy: number): Rect {
    return new Rect(this.x0 + dx, this.y0 + dy, this.x1 + dx, this.y1 + dy);
  }

  /**
   * Scale by factor
   */
  scale(sx: number, sy: number = sx): Rect {
    return new Rect(this.x0 * sx, this.y0 * sy, this.x1 * sx, this.y1 * sy);
  }

  toString(): string {
    return `Rect(${this.x0}, ${this.y0}, ${this.x1}, ${this.y1})`;
  }
}

/**
 * A 2D transformation matrix
 */
export class Matrix implements MatrixLike {
  readonly a: number;
  readonly b: number;
  readonly c: number;
  readonly d: number;
  readonly e: number;
  readonly f: number;

  constructor(a: number, b: number, c: number, d: number, e: number, f: number) {
    this.a = a;
    this.b = b;
    this.c = c;
    this.d = d;
    this.e = e;
    this.f = f;
  }

  /**
   * Identity matrix
   */
  static readonly IDENTITY = new Matrix(1, 0, 0, 1, 0, 0);

  /**
   * Create a matrix from a matrix-like object
   */
  static from(m: MatrixLike): Matrix {
    if (m instanceof Matrix) {
      return m;
    }
    return new Matrix(m.a, m.b, m.c, m.d, m.e, m.f);
  }

  /**
   * Create a translation matrix
   */
  static translate(tx: number, ty: number): Matrix {
    const result = native_addon.matrixTranslate(tx, ty);
    return new Matrix(result.a, result.b, result.c, result.d, result.e, result.f);
  }

  /**
   * Create a scaling matrix
   */
  static scale(sx: number, sy: number = sx): Matrix {
    const result = native_addon.matrixScale(sx, sy);
    return new Matrix(result.a, result.b, result.c, result.d, result.e, result.f);
  }

  /**
   * Create a rotation matrix
   */
  static rotate(degrees: number): Matrix {
    const result = native_addon.matrixRotate(degrees);
    return new Matrix(result.a, result.b, result.c, result.d, result.e, result.f);
  }

  /**
   * Concatenate with another matrix
   */
  concat(other: MatrixLike): Matrix {
    const result = native_addon.matrixConcat(
      { a: this.a, b: this.b, c: this.c, d: this.d, e: this.e, f: this.f },
      { a: other.a, b: other.b, c: other.c, d: other.d, e: other.e, f: other.f }
    );
    return new Matrix(result.a, result.b, result.c, result.d, result.e, result.f);
  }

  /**
   * Pre-translate this matrix
   */
  preTranslate(tx: number, ty: number): Matrix {
    return Matrix.translate(tx, ty).concat(this);
  }

  /**
   * Post-translate this matrix
   */
  postTranslate(tx: number, ty: number): Matrix {
    return this.concat(Matrix.translate(tx, ty));
  }

  /**
   * Pre-scale this matrix
   */
  preScale(sx: number, sy: number = sx): Matrix {
    return Matrix.scale(sx, sy).concat(this);
  }

  /**
   * Post-scale this matrix
   */
  postScale(sx: number, sy: number = sx): Matrix {
    return this.concat(Matrix.scale(sx, sy));
  }

  /**
   * Pre-rotate this matrix
   */
  preRotate(degrees: number): Matrix {
    return Matrix.rotate(degrees).concat(this);
  }

  /**
   * Post-rotate this matrix
   */
  postRotate(degrees: number): Matrix {
    return this.concat(Matrix.rotate(degrees));
  }

  toString(): string {
    return `Matrix(${this.a}, ${this.b}, ${this.c}, ${this.d}, ${this.e}, ${this.f})`;
  }
}

/**
 * A quadrilateral defined by four corner points
 */
export class Quad implements QuadLike {
  readonly ul: Point;
  readonly ur: Point;
  readonly ll: Point;
  readonly lr: Point;

  constructor(ul: PointLike, ur: PointLike, ll: PointLike, lr: PointLike) {
    this.ul = Point.from(ul);
    this.ur = Point.from(ur);
    this.ll = Point.from(ll);
    this.lr = Point.from(lr);
  }

  /**
   * Create a quad from a rectangle
   */
  static fromRect(r: RectLike): Quad {
    return new Quad(
      new Point(r.x0, r.y0),
      new Point(r.x1, r.y0),
      new Point(r.x0, r.y1),
      new Point(r.x1, r.y1)
    );
  }

  /**
   * Transform this quad by a matrix
   */
  transform(m: MatrixLike): Quad {
    const matrix = Matrix.from(m);
    return new Quad(
      this.ul.transform(matrix),
      this.ur.transform(matrix),
      this.ll.transform(matrix),
      this.lr.transform(matrix)
    );
  }

  /**
   * Get the bounding rectangle
   */
  get bounds(): Rect {
    let rect = Rect.EMPTY;
    rect = rect.includePoint(this.ul);
    rect = rect.includePoint(this.ur);
    rect = rect.includePoint(this.ll);
    rect = rect.includePoint(this.lr);
    return rect;
  }

  toString(): string {
    return `Quad(${this.ul}, ${this.ur}, ${this.ll}, ${this.lr})`;
  }
}

File: nanopdf-rs/src/fitz/geometry.rs
//! Geometry primitives - Point, Rect, Matrix, Quad

use std::f32::{INFINITY, NEG_INFINITY};

#[derive(Debug, Clone, Copy, PartialEq, Default)]
pub struct Point { pub x: f32, pub y: f32 }

impl Point {
    pub const ORIGIN: Point = Point { x: 0.0, y: 0.0 };
    pub fn new(x: f32, y: f32) -> Self { Self { x, y } }
    pub fn transform(&self, m: &Matrix) -> Self {
        Self {
            x: self.x * m.a + self.y * m.c + m.e,
            y: self.x * m.b + self.y * m.d + m.f,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Default)]
pub struct Rect { pub x0: f32, pub y0: f32, pub x1: f32, pub y1: f32 }

impl Rect {
    pub const EMPTY: Rect = Rect { x0: INFINITY, y0: INFINITY, x1: NEG_INFINITY, y1: NEG_INFINITY };
    pub const INFINITE: Rect = Rect { x0: NEG_INFINITY, y0: NEG_INFINITY, x1: INFINITY, y1: INFINITY };
    pub const UNIT: Rect = Rect { x0: 0.0, y0: 0.0, x1: 1.0, y1: 1.0 };

    pub fn new(x0: f32, y0: f32, x1: f32, y1: f32) -> Self { Self { x0, y0, x1, y1 } }
    pub fn width(&self) -> f32 { self.x1 - self.x0 }
    pub fn height(&self) -> f32 { self.y1 - self.y0 }
    pub fn is_empty(&self) -> bool { self.x0 >= self.x1 || self.y0 >= self.y1 }
    pub fn is_infinite(&self) -> bool { self.x0 == NEG_INFINITY }
    pub fn contains(&self, x: f32, y: f32) -> bool {
        x >= self.x0 && x < self.x1 && y >= self.y0 && y < self.y1
    }
    pub fn union(&self, other: &Rect) -> Rect {
        Rect {
            x0: self.x0.min(other.x0), y0: self.y0.min(other.y0),
            x1: self.x1.max(other.x1), y1: self.y1.max(other.y1),
        }
    }
    pub fn intersect(&self, other: &Rect) -> Rect {
        Rect {
            x0: self.x0.max(other.x0), y0: self.y0.max(other.y0),
            x1: self.x1.min(other.x1), y1: self.y1.min(other.y1),
        }
    }
    pub fn include_point(&mut self, p: Point) {
        self.x0 = self.x0.min(p.x); self.y0 = self.y0.min(p.y);
        self.x1 = self.x1.max(p.x); self.y1 = self.y1.max(p.y);
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct IRect { pub x0: i32, pub y0: i32, pub x1: i32, pub y1: i32 }

impl IRect {
    pub fn new(x0: i32, y0: i32, x1: i32, y1: i32) -> Self { Self { x0, y0, x1, y1 } }
    pub fn width(&self) -> i32 { self.x1 - self.x0 }
    pub fn height(&self) -> i32 { self.y1 - self.y0 }
    pub fn is_empty(&self) -> bool { self.x0 >= self.x1 || self.y0 >= self.y1 }
}

impl From<Rect> for IRect {
    fn from(r: Rect) -> Self {
        IRect { x0: r.x0.floor() as i32, y0: r.y0.floor() as i32,
                x1: r.x1.ceil() as i32, y1: r.y1.ceil() as i32 }
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Matrix { pub a: f32, pub b: f32, pub c: f32, pub d: f32, pub e: f32, pub f: f32 }

impl Default for Matrix {
    fn default() -> Self { Self::IDENTITY }
}

impl Matrix {
    pub const IDENTITY: Matrix = Matrix { a: 1.0, b: 0.0, c: 0.0, d: 1.0, e: 0.0, f: 0.0 };

    pub fn new(a: f32, b: f32, c: f32, d: f32, e: f32, f: f32) -> Self { Self { a, b, c, d, e, f } }
    pub fn translate(tx: f32, ty: f32) -> Self { Self { a: 1.0, b: 0.0, c: 0.0, d: 1.0, e: tx, f: ty } }
    pub fn scale(sx: f32, sy: f32) -> Self { Self { a: sx, b: 0.0, c: 0.0, d: sy, e: 0.0, f: 0.0 } }
    pub fn rotate(degrees: f32) -> Self {
        let rad = degrees * std::f32::consts::PI / 180.0;
        let (s, c) = (rad.sin(), rad.cos());
        Self { a: c, b: s, c: -s, d: c, e: 0.0, f: 0.0 }
    }
    pub fn concat(&self, m: &Matrix) -> Self {
        Self {
            a: self.a * m.a + self.b * m.c,
            b: self.a * m.b + self.b * m.d,
            c: self.c * m.a + self.d * m.c,
            d: self.c * m.b + self.d * m.d,
            e: self.e * m.a + self.f * m.c + m.e,
            f: self.e * m.b + self.f * m.d + m.f,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Default)]
pub struct Quad { pub ul: Point, pub ur: Point, pub ll: Point, pub lr: Point }

impl Quad {
    pub fn from_rect(r: &Rect) -> Self {
        Self {
            ul: Point::new(r.x0, r.y0), ur: Point::new(r.x1, r.y0),
            ll: Point::new(r.x0, r.y1), lr: Point::new(r.x1, r.y1),
        }
    }
    pub fn transform(&self, m: &Matrix) -> Self {
        Self {
            ul: self.ul.transform(m), ur: self.ur.transform(m),
            ll: self.ll.transform(m), lr: self.lr.transform(m),
        }
    }
}

